### PR TITLE
Make visibility of listed dataset appropriate for version curation statuses and user

### PR DESCRIPTION
**Pull request by sfisher**
I had a much bigger ticket but broke it into multiple.  It is this one  https://github.com/CDL-Dryad/dryad-product-roadmap/issues/300

This PR also goes with a corresponding one in the other repo.

Added tests that show it working.

If this turns out to be too slow later with a large number of records, we may need to find a more efficient way to join rather than doing a subquery.

### CODE REVIEW

File: stash_engine/app/models/stash_engine/curation_activity.rb [hunk 1]
```diff
@@ -148,7 +148,7 @@ def email_orcid_invitations
       authors = resource.authors.where.not(author_email: existing_invites).where.not(author_email: nil)
 
       return if authors.length <= 1
-      authors[1..authors.legnth].each do |author|
+      authors[1..authors.length].each do |author|
```
Comment (briri) [inline stash_engine/app/models/stash_engine/curation_activity.rb]: good catch Scott!

File: stash_engine/app/models/stash_engine/identifier.rb [hunk 2]
```diff
@@ -12,6 +12,30 @@ class Identifier < ActiveRecord::Base
...
+    scope :user_viewable, ->(user: nil) do
+      if user.nil?
+        publicly_viewable
+      elsif user.superuser?
+        Identifier.all
+      elsif user.role == 'admin'
+        with_visibility(states: %w[published embargoed], tenant_id: user.tenant_id)
+      else
+        with_visibility(states: %w[published embargoed], user_id: user.id)
+      end
```
Comment (briri) [inline stash_engine/app/models/stash_engine/identifier.rb]: really nice refactor Scott. Its easy to follow what's happening here which can be really hard to do when these active record queries get complex

File: stash_engine/app/models/stash_engine/resource.rb [hunk 3]
```diff
@@ -147,6 +147,29 @@ def create_share
                                                    status: %w[published embargoed] })
     end
 
+    # complicated join & subquery that may be reused to get the last curation state for each resource
+    SUBQUERY_FOR_LATEST_CURATION = 'SELECT resource_id, max(id) as id FROM stash_engine_curation_activities GROUP BY resource_id'.freeze
+    JOIN_FOR_LATEST_CURATION = "INNER JOIN (#{SUBQUERY_FOR_LATEST_CURATION}) subq ON stash_engine_resources.id = subq.resource_id " \
+      'INNER JOIN stash_engine_curation_activities ON subq.id = stash_engine_curation_activities.id'.freeze
```
Comment (briri) [inline stash_engine/app/models/stash_engine/resource.rb]: Not necessary, but you can use the following syntax to make the above SQL a little easier to read:
```ruby
    SUBQUERY_FOR_LATEST_CURATION = <<~SQL
      SELECT resource_id, max(id) as id
      FROM stash_engine_curation_activities
      GROUP BY resource_id
    SQL
```

**Comment by briri** [comment]
Just need to rebase off master. Feel free to merge yourself once the conflicts have been resolved

**Comment by sfisher** [comment]
Thanks Brian.  I changed the string to a heredoc.

Merging now after merging the latest changes back into this branch from master.

**Pull request merged**